### PR TITLE
Adds the ability to focus on a specific virus symptom using chemicals (makes virology actually good)

### PR DESCRIPTION
**Pull request by RikuTheKiller**
## About The Pull Request

This is a big change, any and all code improvements or pieces of advice are heavily apperciated.

This PR adds the ability to focus on a specific symptom or group of symptoms using specific reagents. Every symptom has it's own list of reagents required to focus on it. I may not be able to interpret reactions between like 60 reagents so finding any conflicts is of top priority.

There is also an additional focus type that uses silver to determine which symptom to neuter/remove when adding formaldehyde or synaptizine, respectively. The rounded up amount of silver in the container is what's used to determine the index of what is neutered/removed. If you added 5u of silver to a beaker with a virus that has 6 symptoms and then tried to neuter or remove a symptom from it, it would neuter/remove the fifth symptom. The same would happen all the way down to 4u of silver, after which it would focus the 4th symptom so on and so forth. Silver focusing works the same as normal focusing in terms of reagent removal.

This does NOT affect the normal way of getting a symptom, you can still do that just fine. However, adding specific reagents to the solution before adding the mutating agent itself will cause it to mutate it's focused symptoms instead of randomly generated ones. This still respects the level range of whatever mutating agent you're using and only combs through ordinarily available.

When a virus mutates focused symptoms it can mutate as many symptoms at once as it physically can hold. Focusing does nothing if the virus symptom cap is reached and any extra symptoms will simply be removed from the return value.

Once a focus has been applied the reagent used is simply deleted from the beaker. This prevents blood duplication issues, a bunch of reaction conflicts and doesn't make you have to go out of your way to separate out the focus reagents.

Currently the only way to see the recipes is through code diving or finding a symptom in-game and then looking at it's description, however I do plan on also adding a completely new section to the https://tgstation13.org/wiki/Infections table of symptoms detailing the reagents required for each symptom.

This is going to be a draft until I get GBP from my other PRs. You can rate it and find bugs in the meanwhile.

## Why It's Good For The Game

This is rather obvious. I don't think there's anyone who enjoys wasting hours clicking on beakers and hoping for RNG to be on your side, like holy shit, it's so fucking boring. Now you have a reason to go explore the station or learn more chemistry to get whatever the hell you need for your god viruses. This does NOT affect balance in regards to the symptoms themselves, only the methods in which you acquire them.

## Changelog

:cl:
add: Added a completely new method for getting specific symptoms called "focusing" which is achieved through specific reagents unique to each symptom.
/:cl:

### CODE REVIEW

**Comment by GuillaumePrata** [comment]
What happens with the focus reagent? It stays behind and must be split from the blood or it gets mixed with the blood itself?

Blood cloning being so easy denies any possible depth on how we treat patients that lost blood and make our blood type system just be a waste, would be nice if this doesn't add more to the pile.

**Comment by RikuTheKiller** [comment]
> What happens with the focus reagent? It stays behind and must be split from the blood or it gets mixed with the blood itself?
> 
> Blood cloning being so easy denies any possible depth on how we treat patients that lost blood and make our blood type system just be a waste, would be nice if this doesn't add more to the pile.

Oh it just gets deleted when the focus is applied. Simple as that.

**Comment by Jacquerel** [comment]
I was a little sad to see you say "there is no way of seeing this information in game" and then add "so I am putting it in another place which is also not in the game".
Is there a way you can put this information in the game?

**Comment by RikuTheKiller** [comment]
> I was a little sad to see you say "there is no way of seeing this information in game" and then add "so I am putting it in another place which is also not in the game". Is there a way you can put this information in the game?

Well, see, to even know what every mutation agent does you need to check the wiki. It's redundant to add it in-game at that point. If you have an idea on how to do it, please inform me or just code it yourself IF this is merged. I literally had to go into a multiple hour coding trance to even get to the point where the PR is at right now. (Guh virology code...)

**Comment by Mothblocks** [comment]
> Well, see, to even know what every mutation agent does you need to check the wiki. It's redundant to add it in-game at that point.

This is not an acceptable mindset. The wiki should aide, it should not be required. We have strived to teach the game through the game itself and adding extra complexity without bothering to teach it is something we do our best to avoid. No, wiki books (and equivalent dumping huge blocks of text) do not count.

**Comment by Mothblocks** [comment]
This requires some sort of interactive UI to teach players, such as embedding it through the PANDemic or similar. There should also be a very quick explanation somewhere of how to actually use the feature.

File: code/datums/diseases/advance/advance.dm [hunk 1]
```diff
@@ -194,6 +194,35 @@
...
+					not_focused -= focus
+					if(!reagents_to_consume.Find(focus))
+						reagents_to_consume += focus
+				if(!not_focused.len) // If a chemical has all of it's focuses matched, the if statement will pass, adding it to the pool.
+					possible_focuses += S
+	if(possible_focuses.len + symptoms.len > VIRUS_SYMPTOM_LIMIT) // If there are too many focused symptoms in the pool to add, we simply remove some until we're under the cap again.
+		for(var/i = possible_focuses.len, i > 0, i--) // Guh C-style loop, had to do it because the entire purpose of the loop is to remove things from the list determining it's length.
+			if (possible_focuses.len <= VIRUS_SYMPTOM_LIMIT)
+				break
+			pick_n_take(possible_focuses)
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:221]:
```suggestion
		for(var/_ = 1 to VIRUS_SYMPTOM_LIMIT - possible_focuses.len)
			pick_n_take(possible_focuses)
```

File: code/datums/diseases/advance/advance.dm [hunk 2]
```diff
@@ -194,6 +194,35 @@
...
+/datum/disease/advance/proc/GetFocusedSymptoms(datum/reagents/container, level_min, level_max)
+	. = list() // The currently focused symptoms, if there are any.
+	if (symptoms.len >= VIRUS_SYMPTOM_LIMIT) // If the symptom cap is reached, we simply return an empty list.
+		return
+	var/list/possible_focuses = list()
+	var/list/reagents_to_consume = list()
+	for(var/symp in SSdisease.list_symptoms)
+		var/datum/symptom/S = new symp
+		if(S.naturally_occuring && S.focuses.len && S.level >= level_min && S.level <= level_max) // Among the usual level and natural occur checks, this checks if the symptom can be focused.
+			if(!HasSymptom(S))
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:207]: Early-continue

File: code/datums/diseases/advance/advance.dm [hunk 3]
```diff
@@ -194,6 +194,35 @@
...
 		. += pick_n_take(possible_symptoms)
 
+/// Returns the currently focused symptoms, if there are any. Determined by the reagents in the container. Maximum returned symptom amount is dependent on the symptom limit.
+/datum/disease/advance/proc/GetFocusedSymptoms(datum/reagents/container, level_min, level_max)
+	. = list() // The currently focused symptoms, if there are any.
+	if (symptoms.len >= VIRUS_SYMPTOM_LIMIT) // If the symptom cap is reached, we simply return an empty list.
+		return
+	var/list/possible_focuses = list()
+	var/list/reagents_to_consume = list()
+	for(var/symp in SSdisease.list_symptoms)
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:204]: Do not shorten variable names, this one is not clear--is this the symptom datum or a type? Rename it to `symptom_type`.

File: code/datums/diseases/advance/advance.dm [hunk 4]
```diff
@@ -194,6 +194,35 @@
...
+		var/datum/symptom/S = new symp
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:205]: Do not shorten variable names, rename this to `symptom`.

File: code/datums/diseases/advance/advance.dm [hunk 5]
```diff
@@ -194,6 +194,35 @@
 	for(var/i = 1; number_of >= i && possible_symptoms.len; i++)
 		. += pick_n_take(possible_symptoms)
 
+/// Returns the currently focused symptoms, if there are any. Determined by the reagents in the container. Maximum returned symptom amount is dependent on the symptom limit.
+/datum/disease/advance/proc/GetFocusedSymptoms(datum/reagents/container, level_min, level_max)
+	. = list() // The currently focused symptoms, if there are any.
+	if (symptoms.len >= VIRUS_SYMPTOM_LIMIT) // If the symptom cap is reached, we simply return an empty list.
+		return
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:201]:
```suggestion
		return .
```

File: code/datums/diseases/advance/advance.dm [hunk 6]
```diff
@@ -194,6 +194,35 @@
 	for(var/i = 1; number_of >= i && possible_symptoms.len; i++)
 		. += pick_n_take(possible_symptoms)
 
+/// Returns the currently focused symptoms, if there are any. Determined by the reagents in the container. Maximum returned symptom amount is dependent on the symptom limit.
+/datum/disease/advance/proc/GetFocusedSymptoms(datum/reagents/container, level_min, level_max)
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:198]:
```suggestion
/datum/disease/advance/proc/get_focused_symptoms(datum/reagents/container, level_min, level_max)
```

(Then fix everything)

File: code/datums/diseases/advance/advance.dm [hunk 7]
```diff
@@ -194,6 +194,35 @@
...
+				if(!not_focused.len) // If a chemical has all of it's focuses matched, the if statement will pass, adding it to the pool.
+					possible_focuses += S
+	if(possible_focuses.len + symptoms.len > VIRUS_SYMPTOM_LIMIT) // If there are too many focused symptoms in the pool to add, we simply remove some until we're under the cap again.
+		for(var/i = possible_focuses.len, i > 0, i--) // Guh C-style loop, had to do it because the entire purpose of the loop is to remove things from the list determining it's length.
+			if (possible_focuses.len <= VIRUS_SYMPTOM_LIMIT)
+				break
+			pick_n_take(possible_focuses)
+	for(var/datum/reagent/reagent as anything in reagents_to_consume)
+		container.del_reagent(reagent)
+	. += possible_focuses // Finally, we return the pool of focused symptoms.
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:224]:
```suggestion
	. += possible_focuses // Finally, we return the pool of focused symptoms.
	return .
```

File: code/datums/diseases/advance/advance.dm [hunk 8]
```diff
@@ -194,6 +194,35 @@
...
+				var/list/not_focused = S.focuses
+				for(var/datum/reagent/focus as anything in S.focuses)
+					if(!container.has_reagent(focus))
+						break
+					not_focused -= focus
+					if(!reagents_to_consume.Find(focus))
+						reagents_to_consume += focus
+				if(!not_focused.len) // If a chemical has all of it's focuses matched, the if statement will pass, adding it to the pool.
+					possible_focuses += S
+	if(possible_focuses.len + symptoms.len > VIRUS_SYMPTOM_LIMIT) // If there are too many focused symptoms in the pool to add, we simply remove some until we're under the cap again.
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:217]: Then you can remove this check after my other change

File: code/datums/diseases/advance/advance.dm [hunk 9]
```diff
@@ -305,9 +334,15 @@
...
-// Randomly generate a symptom, has a chance to lose or gain a symptom.
-/datum/disease/advance/proc/Evolve(min_level, max_level, ignore_mutable = FALSE)
-	if(!mutable && !ignore_mutable)
+// Randomly generate a symptom if no symptoms are focused. If symptoms are focused, adds all of them at once.
+/datum/disease/advance/proc/Evolve(datum/reagents/container, min_level, max_level, ignore_mutable = FALSE)
+	if(!container || !mutable && !ignore_mutable)
+		return
+	var/list/focused_symptoms = GetFocusedSymptoms(container, min_level, max_level)
+	if(focused_symptoms.len)
+		for(var/datum/symptom/S as anything in focused_symptoms)
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:343]: Do not shorten variable names. Be clear about what this is.

File: code/datums/diseases/advance/advance.dm [hunk 10]
```diff
@@ -305,9 +334,15 @@
...
+			AddSymptom(S)
+		Refresh(TRUE)
```
Comment (Mothblocks) [inline code/datums/diseases/advance/advance.dm:345]: Named parameter for TRUE.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/diseases/advance/advance.dm (hunk 1 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:221]: What is this black magic? Guh.

File: code/datums/diseases/advance/advance.dm (hunk 2 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:207]: Convert HasSymptom and the other check above it into an early continue?

File: code/datums/diseases/advance/advance.dm (hunk 4 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:205]: Carried over from the funny generation code.

File: code/datums/diseases/advance/advance.dm (hunk 5 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:201]: Oh, woops.

File: code/datums/diseases/advance/advance.dm (hunk 6 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:198]: This is gonna be fun.

File: code/datums/diseases/advance/advance.dm (hunk 7 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:224]: Once again, isn't . returned here automatically anyways? (Similarly to . = ..() at the start of a method.)

File: code/datums/diseases/advance/advance.dm (hunk 8 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:217]: Got it.

File: code/datums/diseases/advance/advance.dm (hunk 9 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:343]: Uh, sure.

File: code/datums/diseases/advance/advance.dm (hunk 10 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:345]: Will do.

File: code/datums/diseases/advance/advance.dm (hunk 3 above)
Comment (RikuTheKiller) [reply code/datums/diseases/advance/advance.dm:204]: Ah, I get it.

**Comment by RikuTheKiller** [comment]
> This requires some sort of interactive UI to teach players, such as embedding it through the PANDemic or similar. There should also be a very quick explanation somewhere of how to actually use the feature.

Ideas on how to implement this please?

**Comment by RikuTheKiller** [comment]
Okay, I've implemented an automatic addition to the description of every symptom, I'll commit it soon. It contains a line that describes what reagents are required to focus on that symptom. (Find a symptom? Now you know how to get it faster next time.)

It also happens to double as a guide on the feature due to how it's worded. Advanced things like using multiple focuses at once has to be experimented with to understand properly and that's kind of intentional.

**Comment by RikuTheKiller** [comment]
If anyone has an idea on how to add an in-game guide for neutering and removal, please do tell. (It's based off of the amount of silver in the beaker before the mutation agent is added.)

File: tgui/packages/tgui/interfaces/Pandemic/Symptom.tsx [hunk 11]
```diff
@@ -15,12 +15,12 @@ export const SymptomDisplay = (props, context) => {
...
     <Section fill title="Symptoms">
       {symptoms.map((symptom) => {
-        const { name, desc, threshold_desc } = symptom;
+        const { name, desc, focus_desc, threshold_desc } = symptom;
         return (
           <Collapsible key={name} title={name}>
             <Stack fill>
               <Stack.Item grow={3}>
-                {desc}
+                {desc + focus_desc}
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/Pandemic/Symptom.tsx:23]: This should probably be its own item in the stack.

**Comment by Mothblocks** [comment]
I don't know how to help there since I still don't understand what this feature does even after reading your post body several times over

**Comment by davethwave** [comment]
From what I get there are two things this guy is trying to accomplish. One is to allow a way to specifically focus on a symptom inside of a virus in order to neuter/remove it. Such as 4u silver + the blood sample and synaptizine will for certain only effect the 4rth symptom in the list the pandemic shows.
The other effect this would have is to narrow down the symptom you would otherwise get from like mutadone by adding an extra reagent to the mix. Such as if you want viral acceleration but your on your last drop of mutadone. You would instead do 1u mutadone + 1u copper. Thus going from a 1 in 10 chance to get it to maybe a 1 in 3 or just get it straight out. 
The only flaw this PR currently has is that it currently doesn't tell a potential viro player that it exists. Thus making it as if it would have never been merged unless you read the fine print on the wiki. Best likely solution is to somehow integrate it into the pandemic. Perhaps as a message shown when it's empty along with what synaptazine and formalehyde does as well. Working as a basic round up as what virology is as well as a link toward the wiki for more info and all that.

**Comment by RikuTheKiller** [comment]
There is now an added description specifically stating how it works at least for reagent focuses.

I plan on heavily changing individual symptom focus reagents tomorrow. A lot of them aren't very relevant to their symptoms.

**Comment by RikuTheKiller** [comment]
A guidebook dedicated to explaining focuses could be a good ides too. (Not just a wikipage dump.)

**Comment by davethwave** [comment]
Easier too. Just a paper next to the pandemic would likely suffice.

**Comment by Mothblocks** [comment]
Keep it extremely simple and to the point and okay

File: tgui/packages/tgui/interfaces/Pandemic/Symptom.tsx (hunk 11 above)
Comment (RikuTheKiller) [reply tgui/packages/tgui/interfaces/Pandemic/Symptom.tsx:23]: Making it a collapsible seems unnecessary considering it's a single line of text. I've also had problems trying to separate the items and more problems trying to get them to be on different lines.

You likely know more javascript and typescript than me, but seriously, I can't do much more. I've already tried multiple times.

**Comment by RikuTheKiller** [comment]
Alright the reagents used for everything are now MUCH more intuitive.

This does mean I used some inverse reagents, though. Not much of a problem though as inverse reagents are tied to knowledge, not difficulty to obtain.

**Comment by Mothblocks** [comment]
This has not garnered maintainer interest in a few weeks. I recommend reopening when you iron it out with someone like Shaps who has their finger on medical to approve it (I'm out of Discord right now but they talk to me sometimes)

**Pull request closed without merging**